numpy/core/src/multiarray/einsum.dispatch.c.src

```diff
@@ -250,244 +233,198 @@ static void
     @type@ *data0 = (@type@ *)dataptr[0];
     @type@ *data1 = (@type@ *)dataptr[1];
     @type@ *data_out = (@type@ *)dataptr[2];
-
-#if EINSUM_USE_SSE1 && @float32@
-    __m128 a, b;
-#elif EINSUM_USE_SSE2 && @float64@
-    __m128d a, b;
-#endif
-
     NPY_EINSUM_DBG_PRINT1("@name@_sum_of_products_contig_two (%d)\n",
                                                             (int)count);
-
-/* This is placed before the main loop to make small counts faster */
-finish_after_unrolled_loop:
-    switch (count) {
-/**begin repeat2
- * #i = 6, 5, 4, 3, 2, 1, 0#
- */
-        case @i@+1:
-            data_out[@i@] = @to@(@from@(data0[@i@]) *
-                                 @from@(data1[@i@]) +
-                                 @from@(data_out[@i@]));
-/**end repeat2**/
-        case 0:
-            return;
-    }
-
-#if EINSUM_USE_SSE1 && @float32@
+#if @NPYV_CHK@ // NPYV check for @type@
     /* Use aligned instructions if possible */
-    if (EINSUM_IS_SSE_ALIGNED(data0) && EINSUM_IS_SSE_ALIGNED(data1) &&
-        EINSUM_IS_SSE_ALIGNED(data_out)) {
-        /* Unroll the loop by 8 */
-        while (count >= 8) {
-            count -= 8;
-
-/**begin repeat2
- * #i = 0, 4#
- */
-            a = _mm_mul_ps(_mm_load_ps(data0+@i@), _mm_load_ps(data1+@i@));
-            b = _mm_add_ps(a, _mm_load_ps(data_out+@i@));
-            _mm_store_ps(data_out+@i@, b);
-/**end repeat2**/
-            data0 += 8;
-            data1 += 8;
-            data_out += 8;
+    #ifndef NPY_HAVE_NEON
+    const int is_aligned = EINSUM_IS_ALIGNED(data0) && EINSUM_IS_ALIGNED(data1) &&
+                           EINSUM_IS_ALIGNED(data_out);
+    #else
+    // ARM/Neon don't have instructions for aligned memory access
+    const int is_aligned = 0;
+    #endif
+    const int vstep = npyv_nlanes_@sfx@;
+
+    /**begin repeat2
+     * #cond = if(is_aligned), else#
+     * #ld = loada, load#
+     * #st = storea, store#
+     */
+    @cond@ {
+    #if @unroll_by@ == 4
+        const int vstepx4 = vstep * 4;
+        for (; count >= vstepx4; count -= vstepx4, data0 += vstepx4, data1 += vstepx4, data_out += vstepx4) {
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
```

Comment:
Why not use a for loop here?

Reply:
I used for loop at the beginning, but I didn't found a **constant** loop example in the repository , so turn to use the normal `begin repeat` that has been widely used, can you give a hint about insist on using for loop?

Reply:
I'm suggesting the for loop simply because it's easier to read. Are you worried the compiler wouldn't unroll it?

Reply:
I think that the unroll task is supported by compiler, I will replace every `/**begin repeat3` with for loop in order to increase readability.

Reply:
@eric-wieser, @Qiyu8, please stay in the safe side and revert from these changes, since we're not sure of what kind of compiler or flags we're dealing, plus that will not gonna make that much difference with readability.

Reply:
I asked this to start discussion, not as a firm request to change it.

> we're not sure of what kind of compiler or flags we're dealing,

I'm not sure it's our problem if users get slower code due to weird or ancient compilers. There's a trade-off between readability and speed, and while this particular case isn't too close to that line, it would be good to have a policy on where that line is.

At any rate, we should be making these decisions based on benchmarks.

Happy to revert this change for now, but let's see what other reviewers think and keep discussing.

Reply:
for the purpose of saving time, Here lists two loop versions:

- Begin Repeat: Faster & less readability
```
            /**begin repeat3
             * #i = 0, 1, 2, 3#
             */
            npyv_@sfx@ a@i@ = npyv_@ld@_@sfx@(data0 + vstep * @i@);
            npyv_@sfx@ b@i@ = npyv_@ld@_@sfx@(data1 + vstep * @i@);
            npyv_@sfx@ c@i@ = npyv_@ld@_@sfx@(data_out + vstep * @i@);
            /**end repeat3**/
            /**begin repeat3
             * #i = 0, 1, 2, 3#
             */
            npyv_@sfx@ abc@i@ = npyv_muladd_@sfx@(a@i@, b@i@, c@i@);
            /**end repeat3**/
            /**begin repeat3
             * #i = 0, 1, 2, 3#
             */
            npyv_@st@_@sfx@(data_out + vstep * @i@, abc@i@);
            /**end repeat3**/
```
- For loop: Slower & good readability
```
          for (int i = 0; i < 4; i++) {
                npyv_@sfx@ a = npyv_@ld@_@sfx@(data0 + vstep * i);
                npyv_@sfx@ b = npyv_@ld@_@sfx@(data1 + vstep * i);
                npyv_@sfx@ c = npyv_@ld@_@sfx@(data_out + vstep * i);
                npyv_@sfx@ abc = npyv_muladd_@sfx@(a, b, c);
                npyv_@st@_@sfx@(data_out + vstep * i, abc);
            }
```

Reply:
@eric-wieser, compilers in many times during unrolling loops generate strange serialized machine code that effects on [Load-Hit-Store](https://en.wikipedia.org/wiki/Load-Hit-Store), so again I demand reverting these changes.

> I'm not sure it's our problem if users get slower code due to weird or ancient compilers. 

I wouldn't put my full trust in compilers even with the newest versions of GCC/CLANG.

> it would be good to have a policy on where that line is.

Simply treat conv_templates as if it was c++/templates.